### PR TITLE
Add rules S6270, S6302 and S6304 from MMF-2907: AWS IAM Policies 

### LINKED ISSUE #3449: Add rule S6270: Policies authorizing public access to resources are security-sensitive
# What

The goal of this rule is to detect Python code using the AWS CDK that creates resource-based policies granting public access.

This has been previously implemented for Python. See [SONARPY-1144](https://sonarsource.atlassian.net/browse/SONARPY-1144).
# What

# Detection logic

## `PolicyStatement`

Policies are defined by classes such as `Policy` or `ManagedPolicy`. These classes accept a `PolicyDocument`, which contains multiple `PolicyStatement`s. This `PolicyStatement` class should be the main focus of this rule.

### Constructor

The [constructor](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_iam.PolicyStatement.html) receives a single PolicyStatementProps parameters. The following checks should be applied to the parameters of the PolicyStatementProps object that the constructor receives

Status | `principals` | `effect`
-- | -- | --
Noncompliant |  <ul><li> `aws_cdk.aws_iam.StarPrincipal`</li><li>`aws_cdk.aws_iam.AnyPrincipal` </li><li>`aws_cdk.aws_iam.ArnPrincipal` with `"*"` as first parameter (arn) | <ul><li>`iam.Effect.ALLOW`</li><li>Not set
Compliant | Anything other than <ul><li> `aws_cdk.aws_iam.StarPrincipal`</li><li>`aws_cdk.aws_iam.AnyPrincipal` </li><li>`aws_cdk.aws_iam.ArnPrincipal` with `"*"` as first parameter (arn) | Any

**Example**
```js
import { aws_iam as iam } from 'aws-cdk-lib'

new iam.PolicyStatement({
    sid: "AllowAnyPrincipal",
    effect: iam.Effect.ALLOW, // secondary location
    actions: ["s3:*"],
    resources: [bucket.arnForObjects("*")],
    principals: [new iam.AnyPrincipal()] // Noncompliant
})

new iam.PolicyStatement({
    sid: "DenyAnyPrincipal",
    effect: iam.Effect.DENY,
    actions: ["s3:*"],
    resources: [bucket.arnForObjects("*")],
    principals: [new iam.AnyPrincipal()] // Compliant
})

new iam.PolicyStatement({
    sid: "AllowAccountRootPrincipal",
    effect: iam.Effect.ALLOW,
    actions: ["s3:*"],
    resources: [bucket.arnForObjects("*")],
    principals: [new iam.AccountRootPrincipal()] // Compliant
})
```

### `fromJson`

Apart from the constructor, the static method `PolicyStatement.fromJson()` may also be used to create a `PolicyStatement`. It takes an object as first parameter, which should be checked as follows (using JSONPath syntax):

Status | `$.Principal` | `$.Effect`
-- | -- | --
Noncompliant | <ul><li>`*` </li><li> `"AWS": ["*"]` (single element)  </li><li>`"AWS": ["999999999999","*"]` (contains `*` amount other elements) | <ul><li>`"Allow"` (case insensitive) </li><li>`iam.Effect.ALLOW`
Compliant | Anything other than <ul><li>`*` </li><li> `"AWS": ["*"]` (single element)  </li><li>`"AWS": ["999999999999","*"]` (contains `*` amount other elements) | Any

**Example**

```js
import { aws_iam as iam } from 'aws-cdk-lib'

new iam.PolicyStatement.fromJson({
    "Sid": "AnyPrincipal",
    "Effect": "Allow", // secondary location
    "Action": ["kms:*"], 
    "Resource": "*",
    "Principal": {  "AWS": [ "*" ] } // Noncompliant
})

new iam.PolicyStatement.fromJson({
    "Sid": "AllowAccountRootPrincipal",
    "Effect": "Allow",
    "Action": ["kms:*"], 
    "Resource": "*",
    "Principal": { "AWS" : new iam.AccountRootPrincipal().arn } // Compliant
})
```

### `fromJson`

`PolicyDocument` also has a static method `PolicyDocument.fromJson()` which will also indirectly create `PolicyStatements`. This method also takes a object, where `$.Statement` contains a `PolicyStatement[]` which is equal to the input of `PolicyStatement.fromJson()`.

The checks applied to `PolicyStatement.fromJson()` must be applied to each element of `$.Statement`.

**Example**

```js
import { aws_iam as iam } from 'aws-cdk-lib'

const policyDocument = {
    "Version": "2012-10-17",
    "Statement": [{
        "Sid": "AllowAnyPrincipal",
        "Effect": "Allow", // secondary location
        "Action": ["kms:*"], 
        "Resource": "*",
        "Principal": {  "AWS": [ "*", ], // Noncompliant
    },
    {
        "Sid": "AccountRootPrincipal",
        "Effect": "Allow",
        "Action": ["kms:*"], 
        "Resource": "*",
        "Principal": { "AWS" : new iam.AccountRootPrincipal().arn } // Compliant
    }]        
}

const policyDocumentFromJson = iam.PolicyDocument.fromJson(policyDocument)
```

## RSPEC

This [RSPEC PR](https://github.com/SonarSource/rspec/pull/1311) contains the message and highlighting guidelines.

### PR DESCRIPTION
**Pull request by francoismora**
Fixes #3449, #3450 and #3451

### CODE REVIEW

File: eslint-bridge/src/linting/eslint/rules/aws-iam-all-privileges.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  ctx: Rule.RuleContext,
+  options: PolicyCheckerOptions,
+) {
+  const properties = getResultOfExpression(ctx, expr);
+  const effect = getSensitiveEffect(properties, ctx, options);
+  const action = getSensitiveAction(properties, options);
+
+  if (effect.isMissing && action) {
+    ctx.report({
+      message: toEncodedMessage(MESSAGES.message),
```
Comment (victor-diez-sonarsource) [inline eslint-bridge/src/linting/eslint/rules/aws-iam-all-privileges.ts:51]: remove toEncodedMessage when there are no secondary locations

File: eslint-bridge/src/linting/eslint/rules/aws-iam-all-resources-accessible.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  const effect = getSensitiveEffect(properties, ctx, options);
+  const resource = getSensitiveResource(properties, options);
+
+  if (isException(properties, options)) {
+    return;
+  }
+
+  if (effect.isMissing && resource) {
+    ctx.report({
+      message: toEncodedMessage(MESSAGES.message),
```
Comment (victor-diez-sonarsource) [inline eslint-bridge/src/linting/eslint/rules/aws-iam-all-resources-accessible.ts:58]: remove toEncodedMessage when there are no secondary locations

File: eslint-bridge/src/linting/eslint/rules/aws-iam-public-access.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  ctx: Rule.RuleContext,
+  options: PolicyCheckerOptions,
+) {
+  const properties = getResultOfExpression(ctx, expr);
+  const effect = getSensitiveEffect(properties, ctx, options);
+  const principal = getSensitivePrincipal(properties, ctx, options);
+
+  if (effect.isMissing && principal) {
+    ctx.report({
+      message: toEncodedMessage(MESSAGES.message),
```
Comment (victor-diez-sonarsource) [inline eslint-bridge/src/linting/eslint/rules/aws-iam-public-access.ts:61]: remove toEncodedMessage when there are no secondary locations

**Review by victor-diez-sonarsource — CHANGES_REQUESTED** [review]

File: eslint-bridge/src/linting/eslint/rules/aws-iam-all-privileges.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+// https://sonarsource.github.io/rspec/#/rspec/S6317/javascript
```
[the remark this answers is not in the document]
Comment (francoismora) [reply eslint-bridge/src/linting/eslint/rules/aws-iam-all-privileges.ts]: Fixed

File: eslint-bridge/src/linting/eslint/rules/aws-iam-all-resources-accessible.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+// https://sonarsource.github.io/rspec/#/rspec/S6317/javascript
```
[the remark this answers is not in the document]
Comment (francoismora) [reply eslint-bridge/src/linting/eslint/rules/aws-iam-all-resources-accessible.ts]: Fixed

File: eslint-bridge/src/linting/eslint/rules/aws-iam-public-access.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+// https://sonarsource.github.io/rspec/#/rspec/S6317/javascript
```
[the remark this answers is not in the document]
Comment (francoismora) [reply eslint-bridge/src/linting/eslint/rules/aws-iam-public-access.ts]: Fixed

File: eslint-bridge/src/linting/eslint/rules/aws-iam-all-resources-accessible.ts (hunk 2 above)
Comment (francoismora) [reply eslint-bridge/src/linting/eslint/rules/aws-iam-all-resources-accessible.ts:58]: `toEncodedMessage()` is required by the comment-based testing framework that will expect secondary locations when the rule has the SONAR_RUNTIME parameter.

File: eslint-bridge/src/linting/eslint/rules/aws-iam-public-access.ts (hunk 3 above)
Comment (francoismora) [reply eslint-bridge/src/linting/eslint/rules/aws-iam-public-access.ts:61]: Same here

**Review by victor-diez-sonarsource — APPROVED** [review]

File: eslint-bridge/src/linting/eslint/rules/aws-iam-all-privileges.ts (hunk 1 above)
Comment (francoismora) [reply eslint-bridge/src/linting/eslint/rules/aws-iam-all-privileges.ts:51]: `toEncodedMessage()` is required by the comment-based testing framework that will expect secondary locations when the rule has the SONAR_RUNTIME parameter.

**Pull request merged**